idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt

```diff
@@ -0,0 +1,107 @@
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.sealedSubClassToObject
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.lang.Language
+import com.intellij.openapi.project.Project
+import com.intellij.psi.JavaPsiFacade
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.JavaElementType
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.jetbrains.kotlin.KtNodeTypes
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+
+class ConvertSealedSubClassToObjectFix : LocalQuickFix {
+
+    override fun getFamilyName() = "Convert Sealed Sub-class to Object"
+
+    companion object {
+        val JAVA_LANG = Language.findLanguageByID("JAVA")
+        val KOTLIN_LANG = Language.findLanguageByID("kotlin")
+    }
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val klass = descriptor.psiElement.getParentOfType<KtClass>(false) ?: return
+
+        changeInstances(klass)
+        changeDeclaration(klass)
+    }
+
+    /**
+     * Changes declaration of class to object.
+     */
+    private fun changeDeclaration(element: KtClass) {
+        val factory = KtPsiFactory(element)
+
+        element.removeOpen()
+        element.changeToObject(factory)
+        element.transformToObject(factory)
+    }
+
+    private fun KtClass.removeOpen() {
+        modifierList?.getModifier(KtTokens.OPEN_KEYWORD)?.delete()
+    }
+
+    private fun KtClass.changeToObject(factory: KtPsiFactory) {
+        getClassOrInterfaceKeyword()?.replace(factory.createExpression(KtTokens.OBJECT_KEYWORD.value))
+        primaryConstructor?.delete()
+    }
+
+    private fun KtClass.transformToObject(factory: KtPsiFactory) {
+        replace(factory.createDeclaration<KtObjectDeclaration>(text))
+    }
+
+    /**
+     * Replace instantiations of the class with links to the singleton instance of the object.
+     */
+    private fun changeInstances(klass: KtClass) {
+        mapReferencesByLanguage(klass)
+            .apply {
+                replaceKotlin(klass)
+                replaceJava(klass)
+            }
+    }
+
+    /**
+     * Map references to this class by language
+     */
+    private fun mapReferencesByLanguage(klass: KtClass) = ReferencesSearch.search(klass)
+        .groupBy({ it.element.language }, { it.element.parent })
+
+    /**
+     * Replace Kotlin instantiations to a straightforward call to the singleton.
+     */
+    private fun Map<Language, List<PsiElement>>.replaceKotlin(klass: KtClass) {
+        val list = this[KOTLIN_LANG] ?: return
+        val singletonCall = KtPsiFactory(klass).buildExpression { appendName(klass.nameAsSafeName) }
+
+        list.filter { it.node.elementType == KtNodeTypes.CALL_EXPRESSION }
+            .forEach { it.replace(singletonCall) }
```

Comment:
It's probably correct, but we should add some tests for: 1) Callable reference usages like `MyClass::myFunction` or `MyClass()::myFunction`; 2) Nested class usages liked `MyClass::NestedClass()` or `MyClass()::NestedClass()`.